tools/query-events/src/main.rs

```diff
@@ -0,0 +1,205 @@
+use std::collections::HashMap;
+
+use clap::Parser;
+use futures::future::try_join_all;
+use itertools::Itertools;
+use tracing::{error, info};
+use tracing_subscriber::{filter::LevelFilter, EnvFilter};
+
+use tendermint::abci::{Event, EventAttribute};
+use tendermint_rpc::{
+    endpoint::block_results,
+    query::{Condition, Operand, Operation, Query},
+    Client, HttpClient, Order, Url,
+};
+
+#[derive(Debug, Parser)]
+struct Opts {
+    /// The URL of the Tendermint node's RPC endpoint
+    #[clap(short, long)]
+    url: Url,
+
+    /// The query against which blocks should be matched
+    query: Query,
+
+    /// The maximum height at which blocks should be queried (optional)
+    #[clap(long)]
+    max_height: Option<u64>,
+
+    /// Which page to get
+    #[clap(long, default_value = "1")]
+    page: u32,
+
+    /// How many results per page
+    #[clap(long, default_value = "10")]
+    per_page: u8,
+
+    /// How to order the results by height
+    #[clap(long, default_value = "desc")]
+    order: Order,
+
+    /// Enable verbose mode, display partially matching events
+    #[clap(short, long)]
+    verbose: bool,
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() {
+    tracing_subscriber::fmt()
+        .with_env_filter(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env_lossy(),
+        )
+        .init();
+
+    match run().await {
+        Ok(()) => info!("SUCCESS"),
+        Err(e) => error!("{e}"),
+    }
+}
+
+type BoxError = Box<dyn std::error::Error>;
+
+async fn run() -> Result<(), BoxError> {
+    let opts = Opts::parse();
+
+    info!("Connecting to {}", opts.url);
+    let client = HttpClient::new(opts.url).unwrap();
+
+    let query = opts.query.clone();
+
+    info!("Searching for blocks");
+    let result = client
+        .block_search(opts.query, opts.page, opts.per_page, opts.order)
+        .await?;
+
+    let heights = result
+        .blocks
+        .iter()
+        .map(|b| b.block.header.height)
+        .filter(|h| opts.max_height.map_or(true, |max| h.value() <= max))
+        .collect_vec();
+
+    info!(
+        "Found {} blocks with partially matching events at heights: {}",
+        result.total_count,
```

Comment:
When using `--max-height` the count may be heigher, for example with same query as in the description and `--max-height 114467`:
```
2022-11-29T13:41:45.145528Z  INFO query_events: Searching for blocks
2022-11-29T13:41:45.788077Z  INFO query_events: Found 2 blocks with partially matching events at heights: 114467
```
```suggestion
        "Found {} blocks with partially matching events at heights: {}",
        heights.len(),
```